sql: support specifying custom index parameters for geospatial indexes

otan opened the pull request:
Release note (sql change): Introduced the `s2_max_level`, `s2_level_mod`
and `s2_max_cells` storage parameters for modifying the S2 parameters
for indexing GEOMETRY and GEOGRAPHY data types in an inverted index.

Release note (sql change): Introduced the `geometry_min_x`,
`geometry_min_y`, `geometry_max_x`, `geometry_max_y` storage parameters
for indexing GEOMETRY data types in an inverted index.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/52800)
<!-- Reviewable:end -->

sumeerbhola reviewed (changes requested):
Thanks for doing this!

> Reviewed 10 of 10 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52800#-:-MEhrQKADKz8ly4I0jbq:bxgavm9)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan)

___
*[pkg/sql/create_index.go, line 116 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52800#-MEhnMFy7pVP3eMp099S:-MEhnMFy7pVP3eMp099T:b-2h4as2) ([raw file](https://github.com/cockroachdb/cockroach/blob/21e8991e47abcc322ce58cfad0c2bb8c9475b1e2/pkg/sql/create_index.go#L116)):*
> ```Go
> 		return a.applyS2ConfigSetting(evalCtx, key, expr, 0, 30)
> 	case `s2_level_mod`:
> 		return a.applyS2ConfigSetting(evalCtx, key, expr, 1, 30)
> ```

s2.RegionCoverer only supports LevelMod from 1...3.
___
*[pkg/sql/schemaexpr/partial_index.go, line 174 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52800#-MEhqepr4UZFklUxIcU3:-MEhqepr4UZFklUxIcU4:b-giusep) ([raw file](https://github.com/cockroachdb/cockroach/blob/21e8991e47abcc322ce58cfad0c2bb8c9475b1e2/pkg/sql/schemaexpr/partial_index.go#L174)):*
> ```Go
> 
> 		if index.GeoConfig.S2Geometry != nil {
> 			if len(index.ColumnIDs) != 1 {
> ```

why should there be only 1 columnID? The first columnID is the inverted column, but there are also the PK columns. 


<!-- Sent from Reviewable.io -->

otan committed:
sql: support specifying custom index parameters for geospatial indexes

Release note (sql change): Introduced the `s2_max_level`, `s2_level_mod`
and `s2_max_cells` storage parameters for modifying the S2 parameters
for indexing GEOMETRY and GEOGRAPHY data types in an inverted index.

Release note (sql change): Introduced the `geometry_min_x`,
`geometry_min_y`, `geometry_max_x`, `geometry_max_y` storage parameters
for indexing GEOMETRY data types in an inverted index.

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52800)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @sumeerbhola)

___
*[pkg/sql/create_index.go, line 116 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52800#-MEhnMFy7pVP3eMp099S:-MEhvWRvAS6IexN5W8ve:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/21e8991e47abcc322ce58cfad0c2bb8c9475b1e2/pkg/sql/create_index.go#L116)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

s2.RegionCoverer only supports LevelMod from 1...3.
</blockquote></details>

Fixed.
___
*[pkg/sql/schemaexpr/partial_index.go, line 174 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52800#-MEhqepr4UZFklUxIcU3:-MEhvRyC8Lrfe7TxYIuR:b4hci6y) ([raw file](https://github.com/cockroachdb/cockroach/blob/21e8991e47abcc322ce58cfad0c2bb8c9475b1e2/pkg/sql/schemaexpr/partial_index.go#L174)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

why should there be only 1 columnID? The first columnID is the inverted column, but there are also the PK columns. 
</blockquote></details>

I think this is the way it is stored on the descriptor -- in which case this is the only column.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52800#-:-MEy6zt-1hXF22moZH2i:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @sumeerbhola)



<!-- Sent from Reviewable.io -->

otan commented:
bors r=sumeerbhola
